{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/rsrch5/home/plm/phacosta/.local/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n",
      "/usr/local/lib/python3.10/dist-packages/transformers/utils/generic.py:441: UserWarning: torch.utils._pytree._register_pytree_node is deprecated. Please use torch.utils._pytree.register_pytree_node instead.\n",
      "  _torch_pytree._register_pytree_node(\n",
      "2024-12-11 23:38:01.233520: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:479] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered\n",
      "2024-12-11 23:38:01.273435: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:10575] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered\n",
      "2024-12-11 23:38:01.273628: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1442] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n",
      "2024-12-11 23:38:01.326183: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.\n",
      "To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
      "2024-12-11 23:38:02.061414: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT\n",
      "/usr/local/lib/python3.10/dist-packages/transformers/utils/generic.py:309: UserWarning: torch.utils._pytree._register_pytree_node is deprecated. Please use torch.utils._pytree.register_pytree_node instead.\n",
      "  _torch_pytree._register_pytree_node(\n",
      "/usr/local/lib/python3.10/dist-packages/transformers/utils/generic.py:309: UserWarning: torch.utils._pytree._register_pytree_node is deprecated. Please use torch.utils._pytree.register_pytree_node instead.\n",
      "  _torch_pytree._register_pytree_node(\n"
     ]
    }
   ],
   "source": [
    "import os \n",
    "cwd = os.getcwd()\n",
    "os.chdir(cwd.rsplit(\"/\",1)[0])\n",
    "from geneformer import TranscriptomeTokenizer\n",
    "from geneformer import EmbExtractor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"fine_tune\" # fine_tune or pretrained"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cancer = \"lung\"\n",
    "xenium_folder_dict = {\"breast\": \"Xenium_Prime_Human_Lung_Cancer_FFPE_outs\",\n",
    "                      \"lung\":\"Xenium_Prime_Human_Lung_Cancer_FFPE_outs\",\n",
    "                      \"lymph_node\": \"Xenium_Prime_Human_Lymph_Node_Reactive_FFPE_outs\",\n",
    "                      \"prostate\": \"Xenium_Prime_Human_Prostate_FFPE_outs\",\n",
    "                      \"skin\": \"Xenium_Prime_Human_Skin_FFPE_outs\",\n",
    "                      \"ovarian\": \"Xenium_Prime_Ovarian_Cancer_FFPE_outs\"\n",
    "                      }\n",
    "\n",
    "xenium_folder = xenium_folder_dict[cancer]\n",
    "\n",
    "data_path = f\"/rsrch9/home/plm/idso_fa1_pathology/TIER1/paul-xenium/public_data/10x_genomics/{xenium_folder}\"\n",
    "\n",
    "# Use x_scaled and y_scaled as custom attributes\n",
    "custom_attributes = {\n",
    "    \"x_scaled\": \"x_scaled\",\n",
    "    \"y_scaled\": \"y_scaled\",\n",
    "    \"total_counts\": \"total_counts\",  # Include normalized transcript counts\n",
    "    \"target\" : \"class\"\n",
    "}\n",
    "tk = TranscriptomeTokenizer(custom_attributes, nproc=16)\n",
    "\n",
    "# Run the tokenizer\n",
    "tk.tokenize_data(\n",
    "    data_path,\n",
    "    os.path.join(data_path, \"tokenize_output\"),\n",
    "    \"processed_xenium\",\n",
    "    file_format=\"h5ad\"\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "embex = EmbExtractor(model_type=\"Pretrained\",\n",
    "                     emb_layer=-1,\n",
    "                     forward_batch_size=100,\n",
    "                     max_ncells=None,\n",
    "                     nproc=16)\n",
    "\n",
    "# extracts embedding from input data\n",
    "# input data is tokenized rank value encodings generated by Geneformer tokenizer (see tokenizing_scRNAseq_data.ipynb)\n",
    " \n",
    "embs = embex.extract_embs(\"gf-12L-95M-i4096\",\n",
    "                          os.path.join(data_path, \"tokenize_output\", \"processed_xenium.dataset\"),\n",
    "                          os.path.join(data_path, \"embeddings_output\"),\n",
    "                          \"processed_xenium\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
