{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os \n",
    "cwd = os.getcwd()\n",
    "os.chdir(cwd.rsplit(\"/\",1)[0])\n",
    "from geneformer import TranscriptomeTokenizer\n",
    "from geneformer import EmbExtractor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cancer = \"lymph_node\"\n",
    "xenium_folder_dict = {\"breast\": \"Xenium_Prime_Human_Lung_Cancer_FFPE_outs\",\n",
    "                      \"lung\":\"Xenium_Prime_Breast_Cancer_FFPE_outs\",\n",
    "                      \"lymph_node\": \"Xenium_Prime_Human_Lymph_Node_Reactive_FFPE_outs\",\n",
    "                      \"prostate\": \"Xenium_Prime_Human_Prostate_FFPE_outs\",\n",
    "                      \"skin\": \"Xenium_Prime_Human_Skin_FFPE_outs\",\n",
    "                      \"ovarian\": \"Xenium_Prime_Ovarian_Cancer_FFPE_outs\"\n",
    "                      }\n",
    "\n",
    "xenium_folder = xenium_folder_dict[cancer]\n",
    "\n",
    "data_path = f\"/rsrch9/home/plm/idso_fa1_pathology/TIER1/paul-xenium/public_data/10x_genomics/{xenium_folder}\"\n",
    "\n",
    "# Use x_scaled and y_scaled as custom attributes\n",
    "custom_attributes = {\n",
    "    \"x_scaled\": \"x_scaled\",\n",
    "    \"y_scaled\": \"y_scaled\",\n",
    "    \"total_counts\": \"total_counts\"  # Include normalized transcript counts\n",
    "}\n",
    "tk = TranscriptomeTokenizer(custom_attributes, nproc=16)\n",
    "\n",
    "# Run the tokenizer\n",
    "tk.tokenize_data(\n",
    "    data_path,\n",
    "    os.path.join(data_path, \"tokenize_output\"),\n",
    "    \"processed_xenium\",\n",
    "    file_format=\"h5ad\"\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "embex = EmbExtractor(model_type=\"Pretrained\",\n",
    "                     emb_layer=-1,\n",
    "                     forward_batch_size=100,\n",
    "                     max_ncells=None,\n",
    "                     nproc=16)\n",
    "\n",
    "# extracts embedding from input data\n",
    "# input data is tokenized rank value encodings generated by Geneformer tokenizer (see tokenizing_scRNAseq_data.ipynb)\n",
    " \n",
    "embs = embex.extract_embs(\"gf-12L-95M-i4096\",\n",
    "                          os.path.join(data_path, \"tokenize_output\", \"processed_xenium.dataset\"),\n",
    "                          os.path.join(data_path, \"embeddings_output\"),\n",
    "                          \"processed_xenium\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
